### PR TITLE
[WOR-231] Graceful error messages for spend report

### PR DESCRIPTION
**Pull request by ghost**
https://broadworkbench.atlassian.net/browse/WOR-231
Screen Shot 2022-09-16 at 12 41 58 PM
Screen Shot 2022-09-16 at 12 56 22 PM

<!--
Hello, friend!
Remember to mention what sort of testing/verification you performed in the course of building this PR, i.e. checking functionality in the browser locally.
Also, so if a screen recording [1] and/or screenshots would be a helpful way to communicate context, please consider including some in your PR description.
Thanks!

[1] https://support.apple.com/guide/quicktime-player/record-your-screen-qtp97b08e666/mac
--!>

### CODE REVIEW

File: src/pages/billing/Project.js [hunk 1]
```diff
@@ -519,6 +523,14 @@ const ProjectDetail = ({ authorizeAndLoadAccounts, billingAccounts, billingProje
 
   const isGcpProject = billingProject.cloudPlatform === cloudProviders.gcp.label
 
+  const json = Utils.maybeParseJSON(errorMessage)
```
Comment (nawatts) [inline src/pages/billing/Project.js]: This variable could use a more descriptive name. Maybe `error`?

File: src/pages/billing/Project.js [hunk 2]
```diff
@@ -519,6 +523,14 @@ const ProjectDetail = ({ authorizeAndLoadAccounts, billingAccounts, billingProje
...
+  const jsonFrame = {
```
Comment (nawatts) [inline src/pages/billing/Project.js]: Since this is only used once, could it be inlined instead of stored in a variable?

File: src/pages/billing/Project.js [hunk 3]
```diff
@@ -716,8 +754,17 @@ const ProjectDetail = ({ authorizeAndLoadAccounts, billingAccounts, billingProje
         setCostPerWorkspace(costPerWorkspace)
         setUpdatingProjectCost(false)
       }
-    })
-    maybeLoadProjectCost()
+    }
+    maybeLoadProjectCost().catch(async error => {
+      if (error instanceof Response && error.status === 401) {
+        if (!await reloadAuthToken()) {
```
Comment (nawatts) [inline src/pages/billing/Project.js:764]: Can we keep auth related code out of this module and avoid duplicating this code by using one of the standard error handling wrappers?

**Review by nawatts — DISMISSED** [review]
Requesting changes over the error handling code for expired auth tokens. If possible, let's avoid duplicating that in this module.

File: src/pages/billing/Project.js (hunk 3 above)
Comment (cahrens) [reply src/pages/billing/Project.js:764]: I think `withErrorHandling` is probably what you want. Or I think the `catch` block works if you simply delete the 401-specific code. 

@nawatts were you thinking of something else? I know you often just do try/catch.

File: src/pages/billing/Project.js [hunk 4]
```diff
@@ -568,6 +580,34 @@ const ProjectDetail = ({ authorizeAndLoadAccounts, billingAccounts, billingProje
...
     ]),
     [spendReportKey]: div({ style: { display: 'grid', rowGap: '0.5rem' } }, [
+      !!errorMessage && div({
+        style: {
+          backgroundColor: colors.danger(0.15), borderRadius: '4px',
+          boxShadow: '0 0 4px 0 rgba(0,0,0,0.5)', display: 'flex',
+          padding: '1rem', margin: '1rem 0 0'
+        }
+      }, [
+        div({ style: { display: 'flex' }, role: 'alert' },
```
Comment (cahrens) [inline src/pages/billing/Project.js]: Good call on the alert role. However, having it at the top level causes everything to be read out as you expand/collapse the details button. I think it would be sufficient to put the `alert` role just on the `div` that has the shortened message-- so lines 601 and 608.

File: src/pages/billing/Project.js (hunk 3 above)
Comment (nawatts) [reply src/pages/billing/Project.js:764]: Looking at it again, the code for handling expired auth tokens is in `reportError` and this doesn't report the error with a notification. That's probably the wrong place for that code, but that's beside the point. I'll retract my objection.

https://github.com/DataBiosphere/terra-ui/blob/f1c195a9e58f40d5f756df794147210d6be86cb6/src/libs/error.js#L6-L16

Comment (cahrens) [reply src/pages/billing/Project.js:764]: So the point of having the 401 handling is to avoid showing an error about the session timed out inline in the spend report page itself? That's probably a good thing, but if the reload of the Auth token is successfully we would still need to re-execute the Ajax call to load the spend report data, wouldn't we?

In fact, I wonder if this pattern is why I often get an empty terra-ui page when I first bring up Terra-- the auth token had timed out and it refreshed the token, but it still doesn't load any of the actual page content.

Comment (nawatts) [reply src/pages/billing/Project.js:764]: Yeah, I think the existing handling for expired auth tokens is broken in multiple ways:
1. It's in `reportError`. Not every Ajax call is wrapped in `withErrorReporting`. It should be in Ajax.
2. As you point out, it doesn't retry the request. So whatever made the call doesn't work, but now no error message is shown.

Comment (cahrens) [reply src/pages/billing/Project.js:764]: I have created https://broadworkbench.atlassian.net/browse/SUP-743 to fix the underlying issue with auth token expiration handling. @yunip20 for now I think it is fine to keep this code in your PR, and it can be cleaned up when there is more complete handling that retries the original Ajax call.

File: src/pages/billing/Project.js [hunk 5]
```diff
@@ -662,10 +702,8 @@ const ProjectDetail = ({ authorizeAndLoadAccounts, billingAccounts, billingProje
   useEffect(() => { StateHistory.update({ projectUsers }) }, [projectUsers])
   // Update cost data only if report date range changes, or if spend report tab was selected.
   useEffect(() => {
-    const maybeLoadProjectCost = _.flow(
-      reportErrorAndRethrow('Unable to retrieve spend report data'),
-      Utils.withBusyState(setUpdating)
-    )(async () => {
+    const maybeLoadProjectCost = async () => {
```
Comment (cahrens) [inline src/pages/billing/Project.js]: I believe you can still use `Utils.withBusyState` to handling the toggling of `setUpdating`.

File: src/pages/billing/Project.js [hunk 6]
```diff
@@ -716,8 +754,18 @@ const ProjectDetail = ({ authorizeAndLoadAccounts, billingAccounts, billingProje
         setCostPerWorkspace(costPerWorkspace)
         setUpdatingProjectCost(false)
       }
+    }
+    Utils.withBusyState(setUpdating)
```
Comment (ghost) [inline src/pages/billing/Project.js]: I changed to using `Utils.withBusyState()`, but when I use this, I notice that the loading icon/page doesn’t show up even when the page hasn’t fully loaded yet-- I’m not sure if I'm calling Utils.withBusyState() in the right place.

Comment (nawatts) [reply src/pages/billing/Project.js]: [`Utils.withBusyState(...)`](https://github.com/DataBiosphere/terra-ui/blob/9fa8aca84556034aff19401f7aea4a45c83ece67/src/libs/utils.js#L277-L286) returns a function, so this line isn't doing anything.

Typically, it would be used like:
```js
const doSomething = Utils.withBusyState(setUpdating, async () => {
  ...
})
```

It's also often used with an error handling wrapper like:
```js
```js
const doSomething = _.flow(
  withErrorReporting('...'),
  Utils.withBusyState(setUpdating)
)(async () => {
  ...
})
```

File: src/pages/billing/Project.js [hunk 7]
```diff
@@ -568,6 +574,41 @@ const ProjectDetail = ({ authorizeAndLoadAccounts, billingAccounts, billingProje
       ])
     ]),
     [spendReportKey]: div({ style: { display: 'grid', rowGap: '0.5rem' } }, [
+      !!errorMessage && div({
```
Comment (nawatts) [inline src/pages/billing/Project.js]: Can this be extracted into a separate component?

Either just the styles, so that this becomes something like:
```js
!!errorMessage && h(ErrorAlert, [
  Utils.cond(...)
])
```

Or the logic for handling string/object errors as well:
```js
!!errorMessage && h(ErrorAlert, { error })
```

**Review by nawatts — CHANGES_REQUESTED** [review]

File: src/pages/billing/Project.js [hunk 8]
```diff
@@ -524,6 +527,43 @@ const ProjectDetail = ({ authorizeAndLoadAccounts, billingAccounts, billingProje
   const isGcpProject = billingProject.cloudPlatform === cloudProviders.gcp.label
 
   const error = Utils.maybeParseJSON(errorMessage)
+  const ErrorAlert = ({ error }) => {
```
Comment (nawatts) [inline src/pages/billing/Project.js]: `ErrorAlert` should be defined outside of `ProjectDetail`. Components should not be defined inside other components. Because a new `ErrorAlert` function is created each time `ProjectDetail` is rendered, React can't tell that they are the same component. This results in `ErrorAlert` being unmounted and recreated on each render.

**Review by nawatts — CHANGES_REQUESTED** [review]

File: src/pages/billing/Project.js (hunk 8 above)
Comment (ghost) [reply src/pages/billing/Project.js]: Got it. That makes sense, thanks for explaining!

File: src/pages/billing/Project.js [hunk 9]
```diff
@@ -568,6 +615,7 @@ const ProjectDetail = ({ authorizeAndLoadAccounts, billingAccounts, billingProje
       ])
     ]),
     [spendReportKey]: div({ style: { display: 'grid', rowGap: '0.5rem' } }, [
+      !!errorMessage && h(ErrorAlert, { error, errorMessage }),
```
Comment (nawatts) [inline src/pages/billing/Project.js]: It looks like this is the only place `error` is used, so `ErrorAlert` could take one prop (`errorMessage`) and parse it itself.

File: src/pages/billing/Project.js [hunk 10]
```diff
@@ -731,8 +786,12 @@ const ProjectDetail = ({ authorizeAndLoadAccounts, billingAccounts, billingProje
...
   return h(Fragment, [
     div({ style: { padding: '1.5rem 0 0', flexGrow: 1, display: 'flex', flexDirection: 'column' } }, [
-      div({ style: { color: colors.dark(), fontSize: 18, fontWeight: 600, display: 'flex', alignItems: 'center', marginLeft: '1rem' } }, [billingProject.projectName]),
-      isGcpProject && h(GcpBillingAccountControls, { authorizeAndLoadAccounts, billingAccounts, billingProject, isOwner, getShowBillingModal, setShowBillingModal, reloadBillingProject, setUpdating }),
+      div({ style: { color: colors.dark(), fontSize: 18, fontWeight: 600, display: 'flex', alignItems: 'center', marginLeft: '1rem' } },
+        [billingProject.projectName]),
+      isGcpProject && h(GcpBillingAccountControls, {
+        authorizeAndLoadAccounts, billingAccounts, billingProject, isOwner, getShowBillingModal, setShowBillingModal, reloadBillingProject,
+        setUpdating
+      }),
```
Comment (nawatts) [inline src/pages/billing/Project.js]: Generally, it would be nice to avoid unrelated formatting changes like this since they add noise to the diff and the version history.

**Review by nawatts — APPROVED** [review]

**Comment by cahrens** [comment]
@yunip20 given all the billing page integration test failures that began with commit [18f5b02](https://github.com/DataBiosphere/terra-ui/pull/3390/commits/18f5b02bb7d30858a3947d46ca54b6d18eb23990), I think your change has impacted the tests. I think the spinner is no longer showing (which would be set by the busy flag) while the data is loading. Let me know if you need help tracking down this issue.

File: integration-tests/tests/billing-projects.js [hunk 11]
```diff
@@ -33,7 +33,7 @@ const billingProjectsPage = (testPage, testUrl) => {
 
     selectProject: (billingProjectName, cloudPlatform = GCP) => {
       const text = cloudPlatform === GCP ? `Google Cloud Platform${billingProjectName}` : `Microsoft Azure${billingProjectName}`
-      return noSpinnersAfter(testPage, { action: () => click(testPage, clickable({ text })) })
+      return click(testPage, clickable({ text }))
```
Comment (cahrens) [inline integration-tests/tests/billing-projects.js:36]: The test failure was occurring because when we click on a project name, a spinner is not showing anymore from the spend report tab. That actually seems like a good thing, since the spend report tab is not selected. When the tab actually is selected, I do see the spinner as the Ajax data is loaded.

I can't figure out why your code change resulted in this change in behavior, but I think we are OK to just change the test so it no longer expects the spinner to show when the project is selected. I ran the test 100 times with "flakes", and didn't encounter any failures.

Even if a spinner does show when the project is selected, the tests should be fine as long as they don't ever assume that data is immediately present. I believe this is case because the tests are always waiting for some text element or control to appear, and it's OK if that takes some time. There was probably no good reason for the test to ever wait for the spinner at that point.

**Review by cahrens — APPROVED** [review]
